packages/block-library/src/gallery/gallery-image.js

```diff
@@ -34,105 +33,93 @@ import {
 
 const isTemporaryImage = ( id, url ) => ! id && isBlobURL( url );
 
-class GalleryImage extends Component {
-	constructor() {
-		super( ...arguments );
-
-		this.onSelectImage = this.onSelectImage.bind( this );
-		this.onSelectCaption = this.onSelectCaption.bind( this );
-		this.onRemoveImage = this.onRemoveImage.bind( this );
-		this.bindContainer = this.bindContainer.bind( this );
-		this.onEdit = this.onEdit.bind( this );
-		this.onSelectImageFromLibrary = this.onSelectImageFromLibrary.bind(
-			this
-		);
-		this.onSelectCustomURL = this.onSelectCustomURL.bind( this );
-		this.state = {
-			captionSelected: false,
-			isEditing: false,
-		};
-	}
-
-	bindContainer( ref ) {
-		this.container = ref;
-	}
-
-	onSelectCaption() {
-		if ( ! this.state.captionSelected ) {
-			this.setState( {
-				captionSelected: true,
+export default function GalleryImage( {
+	url,
+	alt,
+	id,
+	linkTo,
+	link,
+	isFirstItem,
+	isLastItem,
+	isSelected,
+	caption,
+	sizeSlug,
+	onSelect,
+	onRemove,
+	onMoveForward,
+	onMoveBackward,
+	setAttributes,
+	'aria-label': ariaLabel,
+} ) {
+	const [ captionSelected, setCaptionSelected ] = useState( false );
+	const [ isEditing, setIsEditing ] = useState( false );
+
+	const container = useRef( null );
+
+	const image = useSelect(
+		( select ) => {
+			const { getMedia } = select( 'core' );
+
+			return id ? getMedia( parseInt( id, 10 ) ) : null;
+		},
+		[ id ]
+	);
+
+	const { __unstableMarkNextChangeAsNotPersistent } = useDispatch(
+		'core/block-editor'
+	);
+
+	useEffect( () => {
+		if ( image && ! url ) {
+			__unstableMarkNextChangeAsNotPersistent();
+			setAttributes( {
+				url: image.source_url,
+				alt: image.alt_text,
 			} );
 		}
 
-		if ( ! this.props.isSelected ) {
-			this.props.onSelect();
+		// Unselect the caption so when the user selects other image and comeback
+		// the caption is not immediately selected.
+		if ( captionSelected && ! isSelected ) {
+			setCaptionSelected( false );
 		}
-	}
+	}, [ isSelected, image, url, captionSelected ] );
 
-	onSelectImage() {
-		if ( ! this.props.isSelected ) {
-			this.props.onSelect();
+	const onSelectCaption = () => {
```

Comment:
Personally, I prefer named functions, because they're easier to debug. I don't think there's a good reason to assign anonymous functions to a constant here.